List View top padding

rebornix opened the pull request:
Re #178626. Experimenting top padding for list view by tweaking rangemap.

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

sagar1621 commented:
Hello, I noticed that this issue has been solved and all the automated checks are passing. Would it be possible for a maintainer to review this and merge the changes? Thank you!

joaomoreno commented on src/vs/base/browser/ui/list/rangeMap.ts:
```diff
@@ -91,6 +91,21 @@ export class RangeMap {
+
+	set topPadding(topPadding: number) {
+		this._topPadding = topPadding;
+		this._size = this._topPadding + this.groups.reduce((t, g) => t + (g.size * (g.range.end - g.range.start)), 0);
```
Can't we do better here? Size should only change by `topPadding - this._topPadding`, no need to iterate over items.

joaomoreno commented on src/vs/base/browser/ui/list/rangeMap.ts:
```diff
@@ -91,6 +91,21 @@ export class RangeMap {
+	}
+
+	constructor(topPadding?: number) {
+		this._topPadding = topPadding || 0;
```
```suggestion
		this._topPadding = topPadding ?? 0;
```

joaomoreno reviewed (changes requested):
Good job!

One nit. In https://github.com/microsoft/vscode/issues/178626 you mentioned:

> The catch with this approach is that when users navigate in the list view, the list item doesn't always fully reveal itself (it's off by `${top_padding}px`).

Summarizing the bug: _the list doesn't scroll all the way up if the user navigates to the first item with the keyboard_. I actually can't reproduce this in `main` but I can in this PR's changes. Can you clarify exactly what you meant by that and what's your expected behavior?

I would also love the following changes on top of this PR:

- Rename `topPadding` to `paddingTop`
- Rename `additionalScrollHeight` to `paddingBottom`

rebornix commented:
@joaomoreno thanks for the review and didn't rush to get fixes in and merge last iteration. Now I addressed all the comments, the padding is now `paddingTop` and `paddingBottom`.

> Summarizing the bug: the list doesn't scroll all the way up if the user navigates to the first item with the keyboard

It's actually the opposite, when users press arrow down, we fail to reveal the full cell before. Here is a comparison between now (left side) and before (after)

https://github.com/microsoft/vscode/assets/876920/60d2a3f0-6341-4d8a-a794-9ca966c617d6

In this video you can see that, in OSS with this branch, we can fully reveal the cell when navigating down, allowing users to see the full output along with the "insert toolbar", while on the right side, it didn't reveal enough, and often the output can be cut off and also we don't have enough space for the insert toolbar.

For scrolling up, the experience you are seeing is then the same as text editor in below video: it only reveals the top of the first cell, but it won't scroll all the way to the top padding.

https://github.com/microsoft/vscode/assets/876920/48c3487e-c610-4108-a77e-f6820a2687a3
